### ISSUE TITLE
☂️ [GEP-20] Highly Available Seed and Shoot Clusters

### ISSUE THREAD

**Issue by shreyas-s-rao** [gardener/gardener#6529]
**How to categorize this issue?**

/area high-availability
/kind enhancement

**What would you like to be added**:

This is an umbrella issue to track the implementation of [GEP-20 Highly Available Shoot Control Planes](https://github.com/gardener/gardener/blob/master/docs/proposals/20-ha-control-planes.md).

## Tasks

- [x] https://github.com/gardener/gardener/pull/6530
- Add validations for updating the `shoot.spec.controlPlanes` field
  - [x] Allow non-HA shoot -> HA shoot
  - [x] Only allow non-HA -> multi-zone if assigned seed is multi-zonal
  - [x] Single-zone HA shoot <-> multi-zone HA shoot must not be allowed
  - [x] HA shoot -> non-HA shoot must not be allowed (until etcd scale-down is implemented)
- [x] Enhance validations for shoot HA annotation
- [x] https://github.com/gardener/gardener/pull/6723 
  - [x] https://github.com/gardener/gardener/pull/6914
  - [x] https://github.com/gardener/gardener/pull/6960
- [x] https://github.com/gardener/gardener/pull/6579
  - [x] https://github.com/gardener/gardener/pull/6934
- [x] ~~Enhance Pod eviction in case of zone outage (delete Pods in `Terminating` state), see https://github.com/gardener/gardener/pull/6646#issuecomment-1243536333~~ (@ialidzhikov) -> see https://github.com/gardener/gardener/issues/6529#issuecomment-1254800255
- [x] https://github.com/gardener/gardener/pull/6665
- Configure high-availability settings for {seed system, shoot control plane, shoot system} components
  - [x] https://github.com/gardener/gardener/pull/6750 
  - [x] https://github.com/gardener/gardener/pull/6685
  - [x] https://github.com/gardener/gardener/pull/6967
    - [x] https://github.com/gardener/gardener/pull/6982
    - [x] https://github.com/gardener/gardener/pull/6989
    - [x] https://github.com/gardener/gardener/pull/6992
  - Remainders/special cases
    - [x] @MartinWeindel: `vpn-seed-server` and `vpn-shoot`, https://github.com/gardener/gardener/issues/6890
      - https://github.com/gardener/gardener/pull/6978
    - [x] @ScheererJ: `istio-ingressgateway`
      - https://github.com/gardener/gardener/pull/6997
    - [x] https://github.com/gardener/gardener/issues/7105
  - Adapt all extensions
    - [x] [provider-openstack](https://github.com/gardener/gardener-extension-provider-openstack): https://github.com/gardener/gardener-extension-provider-openstack/pull/515, released with [`v1.31.0`](https://github.com/gardener/gardener-extension-provider-openstack/releases/tag/v1.31.0)
    - [x] [provider-equinix-metal](https://github.com/gardener/gardener-extension-provider-equinix-metal): https://github.com/gardener/gardener-extension-provider-equinix-metal/pull/237, ~~released with [`v?`](https://github.com/gardener/gardener-extension-provider-equinix-metal/releases/tag/v?)~~ (no need to wait for it)
    - [x] [provider-vsphere](https://github.com/gardener/gardener-extension-provider-vsphere): https://github.com/gardener/gardener-extension-provider-vsphere/pull/321, released with [`v1.23.0`](https://github.com/gardener/gardener-extension-provider-vsphere/releases/tag/v1.23.0)
    - [x] [provider-aws](https://github.com/gardener/gardener-extension-provider-aws): https://github.com/gardener/gardener-extension-provider-aws/pull/645, released with [`v1.41.0`](https://github.com/gardener/gardener-extension-provider-aws/releases/tag/v1.41.0)
    - [x] [provider-gcp](https://github.com/gardener/gardener-extension-provider-gcp): https://github.com/gardener/gardener-extension-provider-gcp/pull/521, released with [`v?`](https://github.com/gardener/gardener-extension-provider-gcp/releases/tag/v?)
    - [x] [provider-azure](https://github.com/gardener/gardener-extension-provider-azure): https://github.com/gardener/gardener-extension-provider-azure/pull/599, released with [`v1.33.0`](https://github.com/gardener/gardener-extension-provider-azure/releases/tag/v1.33.0)
    - [x] [provider-alicloud](https://github.com/gardener/gardener-extension-provider-alicloud): https://github.com/gardener/gardener-extension-provider-alicloud/pull/545, released with [`v1.43.0`](https://github.com/gardener/gardener-extension-provider-alicloud/releases/tag/v1.43.0)
    - [x] [networking-cilium](https://github.com/gardener/gardener-extension-networking-cilium): https://github.com/gardener/gardener-extension-networking-cilium/pull/138, released with [`v1.20.0`](https://github.com/gardener/gardener-extension-networking-cilium/releases/tag/v1.20.0)
    - [x] [networking-calico](https://github.com/gardener/gardener-extension-networking-calico): https://github.com/gardener/gardener-extension-networking-calico/pull/222, released with [`v1.28.0`](https://github.com/gardener/gardener-extension-networking-calico/releases/tag/v1.28.0)
    - [x] [shoot-networking-filter](https://github.com/gardener/gardener-extension-shoot-networking-filter): https://github.com/gardener/gardener-extension-shoot-networking-filter/pull/41, released with [`v0.9.0`](https://github.com/gardener/gardener-extension-shoot-networking-filter/releases/tag/v0.9.0)
    - [x] [shoot-networking-problem-detector](https://github.com/gardener/gardener-extension-shoot-networking-problemdetector): https://github.com/gardener/gardener-extension-shoot-networking-problemdetector/pull/37, released with [`v0.9.0`](https://github.com/gardener/gardener-extension-shoot-networking-problemdetector/releases/tag/v0.9.0)
    - [x] [os-ubuntu](https://github.com/gardener/gardener-extension-os-ubuntu): https://github.com/gardener/gardener-extension-os-ubuntu/pull/72, released with [`v1.20.0`](https://github.com/gardener/gardener-extension-os-ubuntu/releases/tag/1.20.0)
    - [x] [os-gardenlinux](https://github.com/gardener/gardener-extension-os-gardenlinux): https://github.com/gardener/gardener-extension-os-gardenlinux/pull/70, released with [`v0.16.0`](https://github.com/gardener/releases/tag/v?/gardener-extension-os-gardenlinux/releases/tag/v0.16.0)
    - [x] [os-suse-chost](https://github.com/gardener/gardener-extension-os-suse-chost): https://github.com/gardener/gardener-extension-os-suse-chost/pull/76, released with [`v1.20.0`](https://github.com/gardener/gardener-extension-os-suse-chost/releases/tag/v1.20.0)
    - [x] [os-coreos](https://github.com/gardener/gardener-extension-os-coreos): https://github.com/gardener/gardener-extension-os-coreos/pull/52, released with [`v1.16.0`](https://github.com/gardener/gardener-extension-os-coreos/releases/tag/v1.16.0)
    - [x] [shoot-dns-service](https://github.com/gardener/gardener-extension-shoot-dns-service): https://github.com/gardener/gardener-extension-shoot-dns-service/pull/170, released with [`v1.28.0`](https://github.com/gardener/releases/tag/v?/gardener-extension-shoot-dns-service/releases/tag/v1.28.0)
    - [x] [shoot-cert-service](https://github.com/gardener/gardener-extension-shoot-cert-service): https://github.com/gardener/gardener-extension-shoot-cert-service/pull/138, released with [`v1.27.0`](https://github.com/gardener/gardener-extension-shoot-cert-service/releases/tag/v1.27.0)
    - [x] [shoot-oidc-service](https://github.com/gardener/gardener-extension-shoot-oidc-service): https://github.com/gardener/gardener-extension-shoot-oidc-service/pull/75, released with [`v0.15.0`](https://github.com/gardener/gardener-extension-shoot-oidc-service/releases/tag/v0.15.0)
    - [x] [runtime-gvisor](https://github.com/gardener/gardener-extension-runtime-gvisor): https://github.com/gardener/gardener-extension-runtime-gvisor/pull/66, released with [`v0.7.0`](https://github.com/gardener/gardener-extension-runtime-gvisor/releases/tag/v0.7.0)
    - [x] [registry-cache](https://github.com/gardener/gardener-extension-registry-cache): https://github.com/gardener/gardener-extension-registry-cache/pull/5, released with [`v0.1.0`](https://github.com/gardener/gardener-extension-registry-cache/releases/tag/v0.1.0)
    - [x] [kupid](https://github.com/gardener/kupid): https://github.com/gardener/kupid/pull/51, released with [`v0.5.0`](https://github.com/gardener/kupid/releases/tag/v0.5.0)
    - [x] ~~auditlog-extension: kubernetes/auditlog-extension#388, released with `v?`~~
- [x] Drop support for `alpha.control-plane.shoot.gardener.cloud/high-availability` annotation
  - https://github.com/gardener/gardener/pull/6998
  - Remove annotation entirely from code after v1.63 got released: https://github.com/gardener/gardener/pull/7493
- [x] https://github.com/gardener/gardener/issues/7688
- [x] https://github.com/gardener/gardener/issues/7689
- [x] https://github.com/gardener/gardener/issues/7690
- [x] https://github.com/gardener/gardener/pull/7695
- [x] Support control-plane migration for HA shoots: https://github.com/gardener/gardener/pull/7626
  - [x] Add e2e test for CP migration of HA Shoot: https://github.com/gardener/gardener/pull/7742
  - [x] Enable the e2e test for CP migration of HA Shoot in ci-infra/prow: https://github.com/gardener/ci-infra/pull/705
- [x] https://github.com/gardener/gardener/issues/7224
- [x] https://github.com/gardener/gardener/pull/7867
- [x] Promote `HAControlPlanes` feature gate to GA: https://github.com/gardener/gardener/pull/8008
- [x] https://github.com/gardener/gardener/pull/8083

**Comment by ashwani2k** [comment]
> - [ ] Introduce shoot spec field for enabling HA control planes
> - Add validations for updating the `shoot.spec.controlPlanes` field
>   - [ ] Allow non-HA shoot -> HA shoot
>   - [ ] Only allow non-HA -> multi-zone if assigned seed is multi-zonal
>   - [ ] Single-zone HA shoot <-> multi-zone HA shoot must not be allowed
>   - [ ] HA shoot -> non-HA shoot must not be allowed (until etcd scale-down is implemented)

This needs some modifications along with a change required for Seed. 
Change needs to be part of GEP -> GEP enhancement PR | Implementation--> Review  

We reprioritised in discussion with @timuthy 
> - [ ] Support zone-pinning for single-zone control HA planes (via GRM mutating webhook)

**Comment by ashwani2k** [comment]
Below bullet points highlight the api contract for introducing HA control planes  via - 
```
  controlPlane:
    highAvailability:
      failureTolerance:
        type:  <node|zone>
```

1. `non-HA` shoots can be scheduled on `non-HA` or` HA`(_multi-zone_) seeds.
2. `single-zone` shoots can be scheduled on `non-HA` or `HA`(_multi-zone_) seeds.
3. `multi-zone` shoots can only be scheduled **ONLY on** HA(multi-zone) seeds.
4. `non-HA` shoots can be upgraded to `single-zone`  on `non-HA` or `HA` seeds. **
5. `non-HA` shoots can be upgraded to `multi-zone` **only on** `HA` seeds. **
6. **`single-zone` shoots shall not be allowed to upgrade to `multi-zone` shoots and shall be stopped by admission plugins.**

**  _this can lead to a short disruption|downtime when `etcd` `sts` is rolled_

---
_Legend_:
**`non-HA` shoot** : _any shoot which has no `faultTolerance`  defined._
**`single-zone` shoot**: _any shoot which has `faultTolerance` defined as type `node`_.
**`multi-zone` shoot**: _any shoot which has `faultTolerance` defined as type `zone`_.
**`non-HA` seed**: _any seed which has worker pools for etcd/cpu running only on a single availability zone_.
**`HA` seed**: _any seed which has worker pools for etcd/cpu defined across 3 availability zones and has the label `seed.gardener.cloud/multi-zonal: "true"`_.

**Comment by ialidzhikov** [comment]
Wrt to `Enhance Pod eviction in case of zone outage (delete Pods in Terminating state)`: for Deployments the kube-controller-manager behaviour is to create new Pods right away when the old Pods are Terminating.

Example:

<details>

<summary>Expand to see the Deployment!</summary>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      terminationGracePeriodSeconds: 60
      containers:
      - name: nginx
        image: centos
        command: ["/bin/sh"]
        args: ["-c", "sleep 3600"]
        ports:
        - containerPort: 80
```

</details>

Above we have a Deployment. Its container does not handle SIGTERM and it will hang in Terminating until it is force killed after `terminationGracePeriodSeconds`.

```
$ k get po
NAME                                READY   STATUS    RESTARTS   AGE
nginx-deployment-746759f465-z95lj   1/1     Running   0          11m

$ k delete po nginx-deployment-746759f465-z95lj
pod "nginx-deployment-746759f465-z95lj" deleted

$ k get po
NAME                                READY   STATUS              RESTARTS   AGE
nginx-deployment-746759f465-z95lj   1/1     Terminating         0          11m
nginx-deployment-746759f465-ztz65   0/1     ContainerCreating   0          2s

$ k get po
NAME                                READY   STATUS        RESTARTS   AGE
nginx-deployment-746759f465-z95lj   1/1     Terminating   0          12m
nginx-deployment-746759f465-ztz65   1/1     Running       0          54s
```

Above you can see that when old replica is deleted, the new one is created right away.

---

I suspect that in the experiments of @unmarshall (ref https://github.com/gardener/gardener/pull/6287#discussion_r942324110) there was a webhook preventing creation of new Pods (for some unknown reason) or `kube-controller-manager` was down (for some unknown reason). These are the 2 potential things that could explain https://github.com/gardener/gardener/pull/6287#discussion_r942324110.

Anyways, I will try to simulate a zone outage and check why KCM does not create the new Pods when the old ones are terminating.

**Comment by ialidzhikov** [comment]
We had a sync @unmarshall and we able to confirm that in a simulation of zone outage (simulated via network acl that denies all ingress and egress traffic for a zone) the recovery for a multi-zone control plane worked as outlined in https://github.com/gardener/gardener/issues/6529#issuecomment-1253517049:
- For Deployments kube-controller-manager creates new replicas right away when the old replicas are terminating. The new replicas start successfully on a healthy zone.
- We also found that the garbage-collector already delete

**Comment by ialidzhikov** [comment]
We had a sync with @unmarshall and we are able to confirm that in a simulation of zone outage (simulated via network acl that denies all ingress and egress traffic for a zone) the recovery for a (multi-zone) control plane worked well as outlined in https://github.com/gardener/gardener/issues/6529#issuecomment-1253517049:
- For Deployments kube-controller-manager creates new replicas right away when the old replicas are terminating. The new replicas start successfully on a healthy zone.
- I think during @unmarshall's simulations kube-controller-manager was down for some reason. I also revised the webhooks we deploy and whether we could have a deadlock situation that could block new Pod creation but I didn't see anything abnormal.

PS: We also found that the existing garbage-collector (shoot-care-controller of gardenlet) already deletes Terminating pods in the Shoot's control plane after `5min`. 
https://github.com/gardener/gardener/blob/24b667c39f4ff1c6e733ce98816e4f6127a2476f/pkg/operation/care/garbage_collection.go#L85-L93
But this is not a recovery mechanism and does not bring to the recovery. For Deployments kube-controller-manager already creates the new replicas. For StatefulSets, even when the old Terminating replicas are forcefully deleted, this does not lead to a recovery as the new StatefulSet Pods fail to be scheduled - they have scheduling requirements that cannot be satisfied during the zone outage (etcd Pod to run on the outage zone or loki/prometheus Pods to run on the outage zone because their volume is already on provisioned on this zone).

TL;DR: We will resolve the corresponding item as completed as nothing has to be done. Let us know if you have additional comments on this topic. We have to update GEP-20 with the new learnings.

**Comment by vlerenc** [comment]
Great to hear! Thank you!

**Comment by timuthy** [comment]
I added another item `Support control-plane migration for HA shoots` since this doesn't seem to work out of the box. We should create a separate issue once we have more certainty and details and find a proper way to support this use-case.

cc @plkokanov @vlerenc

**Comment by plkokanov** [comment]
> I added another item `Support control-plane migration for HA shoots` since this doesn't seem to work out of the box. We should create a separate issue once we have more certainty and details and find a proper way to support this use-case.
> 
> cc @plkokanov @vlerenc

Should we (for now) add validation that forbids migration for HA shoots?

**Comment by rfranzke** [comment]
/assign @rfranzke 

for the tasks related to
> Configure high-availability settings for {seed system, shoot control plane, shoot system} components

(except for the Remainders/special cases)

**Comment by rfranzke** [comment]
> /assign @rfranzke
> 
> for the tasks related to
> 
> > Configure high-availability settings for {seed system, shoot control plane, shoot system} components
> 
> (except for the Remainders/special cases)

All related PRs in `gardener/gardener` have been merged.
/unassign

**Comment by timuthy** [comment]
/assign @plkokanov @ishan16696 
for tasks related to
> Support control-plane migration for HA shoots

**Comment by ishan16696** [comment]
>/assign @plkokanov @ishan16696
for tasks related to

Please see the approaches possible to achieve CPM in multi-node etcd: https://github.com/gardener/etcd-druid/issues/479#issuecomment-1365793557

**Comment by rfranzke** [comment]
All tasks have been completed.
/close
